### ISSUE TITLE
LZMA header decompression fails with error 'Data is corrupt'

### ISSUE THREAD

**Issue by philsmd** [philsmd/7z2hashcat#2]
I (the author of this tool) got some reports over irc and hashcat's forum that in very rare cases with very large .7z files the LZMA decompression of the header sometimes fails.

I think, after a lot of investigation/troubleshooting/specs reading/code debugging (including the code from 7z-SDK, xz tools, p7zip, 7z2hashcat.pl etc), that the main problem might be caused by this small difference of behaviour of the decoders (that I found after re-reading the specs a couple of times, because it was hidden in a small paragraph in DOC/lzma-specification.txt).

The file DOC/lzma-specification.txt of the 7z-SDK says (quote):
"The reference LZMA Decoder ignores the value of the "Corrupted" variable. So it continues to decode the stream, even if the corruption can be detected in the Range Decoder. To provide the full compatibility with output of the reference LZMA Decoder, another LZMA Decoder implementations must also ignore the value of the "Corrupted" variable."

For now, I have improved the warning/error messages a little bit (to include this info), see for instance this commit: https://github.com/philsmd/7z2hashcat/commit/9b919a59e424909efd7acc03ef9a89e061ede350

We might need to find a way to work around this problem (and I guess we need to go the hard way to do so, because de-escalating this 'Data is corrupt' problem, such that it does continue to decompress, might not be possible at all with the available perl modules for LZMA and especially the Compress::Raw::Lzma module currently used).

I might continue to investigate this and update 7z2hashcat.pl accordingly. Any suggestions on how to best fix this issue are very welcome.

---

Update: I totally forgot to include an example. Below is a minimal .7z file that will result in a 'Data is corrupt' warning message when you run it with 7z2hashcat.pl.  

Please note this 7z file is very small in size but this is only the case because it was heavily compressed with LZMA (and contains both easy compressible file names + good compressible data chunks). In fact, it contains about 15000 files with very long file names. It seems very likely that this is exactly a case of this missing 'Data is corrupt' ignoring feature.

I can't exclude that there are better (and even smaller 7z files) that trigger the same error. If you have some very small 7z file that cause the same LZMA decompression warning, please attach it or send it to me. Thx
(just to make it very clear: I ask for very small 7z files, because they are easier to analyze/dissect... it's always a mess if you need to deal with huge "seamingly random" - because compressed - data)

[lzma_data_corrupt.zip](https://github.com/philsmd/7z2hashcat/files/695606/lzma_data_corrupt.zip)
(please rename the file extension of the above downloaded file from .7z to .zip, because it seems github doesn't allow 7z uploads!)

Note: This file isn't encrypted at all (so no password is needed). Of course 7z2hashcat.pl won't output a hash even if it was correctly parsed without an error message. In fact, it should result in a "WARNING: no AES data found in the 7z file 'lzma_data_corrupt.7z'" warning. Fortunately, we do not need encryption (too) to understand this issue and therefore we should get rid of if to solve this problem (otherwise it could potentially be an additional point of failure).

**Comment by philsmd** [comment]
Update: I just pushed some changes and it seems that it fixed some problems with the LZMA decoding code. See: https://github.com/philsmd/7z2hashcat/commit/1f92e4fca1b9a317f45a786dfab99800728646f0

The problem sometimes was that using the lzma2 chunked encoding schema (previously the lzma_generate_header () function was used to generate the header), we had this limits:
$LZMA2_CHUNK_MAX = 1 << 16;
$LZMA2_UNCOMPRESSED_MAX = 1 << 21;

... but sometimes the data from the 7z and therefore the decoder and header information needed more decompressed bytes (7z/xz call it incorrectly "uncompressed" bytes) than 2 ^ 21 = 2097152.
There to data had to be split into multiple chunks.
Since the data extracted from the 7-Zip archives do not contain the data formatted in those chunks I had to switch to a different decoder (from RawDecoder () to AloneDecoder ()).

This allows a much larger uncompressed size encoding + seems to be able to deal with the extracted blob from the archives AS-IS (but with the added new header consisting of Properties (1 byte, defaults to 0x5d), Dictionary Size  (4 bytes little endian), Uncompressed Size (8 bytes little endian).

As you might have noted, now we have a 64 bit uncompressed size field (so the maxium changed from 2097152 to a theoretical max of 18446744073709551615).

Note: when I run 7z2hashcat.pl with the new code on lzma_data_corrupt.zip (or with .7z extension, it doesn't really matter), I get the expected output (note as said above, I intentionally didn't encrypt this file to keep the points of failures low):
"WARNING: no AES data found in the 7z file 'lzma_data_corrupt.zip'"

I'm pretty sure that this fix solved some issues and that now much more .7z files can be correctly parsed.

I will keep this issue open for the time being, because I'm still not sure about what we should think/do about the specs statement above (especially this one: "implementations must also ignore the value of the "Corrupted" variable").
I might close this issue if I do not get any further reports with the fixed version. Furthermore, I would need to generate/have a .7z file that still triggers a decoding error... and fortunately I currently have none such case. That's actually goog. Hurray!

**Comment by syb3ria** [comment]
Hello,
just tried version 0.5 and it gives me the following output:
```
D:\>7z2hashcat.pl Archive.7z
substr outside of string at D:\cRARc\7z2hashcat.pl line 117.
Use of uninitialized value in substr at D:\cRARc\7z2hashcat.pl line 113.
Use of uninitialized value in substr at D:\cRARc\7z2hashcat.pl line 117.
substr outside of string at D:\cRARc\7z2hashcat.pl line 117.
```
Archive.7z, size - 38Gb, Folders - 142, Files -10520, Compression - store.

Regards,

**Comment by philsmd** [comment]
That might imply that the LZMA decompression now works (also for you).
I've opened a new issue for the substr () problem, see https://github.com/philsmd/7z2hashcat/issues/3

We shouldn't mix these problems, because they are very likely completely unrelated.

Please continue to report/give feedback for the LZMA decompression issue on this github issue.
Other problems might/should get their own github issue id.

Anyway, thank you very much for your testings and reportings... this might help a lot to get rid of some bugs and implement some unsupported/new features (and special cases).

**Comment by syb3ria** [comment]
Version 0.6 gives no output at all.

**Comment by philsmd** [comment]
I'm sorry @syb3ria syb3ria but I can't really troubleshoot/debug this without having the file you are using.
You either need to attach it here or send it to me privately (e.g. via mail @ philsmd [a t - s i g n] hashcat DOT net).
That may help to find the root cause of the problem you are experiencing...
I highly doubt that this is still related to the LZMA decompression issue (because otherwise you either would see an error/warning message or get a correct hash output)... so I think we should discuss this "empty output" problem within a new issue.
It would help a lot if you are able to provide the file.

**Comment by syb3ria** [comment]
As the file is too big to be uploaded here (max size 10mb), as soon as I upload it to mega i'll share the download link.

**Comment by syb3ria** [comment]
I was able to reproduce the same scenario with another archive. Here is the download link:

#!wJIjgBha!Z_yRVQYsIhHYwQUOu_ceAhsHhZoBI5L2CCbxL81wTSc

it's at mega.nz

**Comment by philsmd** [comment]
Thank you for this example .7z file. I just noticed that you posted the link here and started the download. Earlier today I fixed some (maybe related) problem and pushed it to github (see https://github.com/philsmd/7z2hashcat/commit/3d25fd186d3e38a06a232f621c0d95c3627f9686).

I still need to verify (after I've downloaded your example) if it fixes also the problem you are experiencing.
I will investigate this carefully soon. Thank you very much
